auth/src/test/java/com/firebase/ui/auth/ui/idp/CredentialSignInHandlerTest.java

```diff
@@ -0,0 +1,191 @@
+        FirebaseUser mockFirebaseUser = TestHelper.makeMockFirebaseUser();
+        IDPResponse mockIdpResponse = mock(IDPResponse.class);
+
+        CredentialSignInHandler credentialSignInHandler = new CredentialSignInHandler(
+                mockActivty,
+                mockActivityHelper,
+                RC_ACCOUNT_LINK,
+                RC_SAVE_CREDENTIALS,
+                mockIdpResponse);
+
+        Context mockContext = mock(Context.class);
+        FlowParameters mockFlowParams = mock(FlowParameters.class);
+        Task mockTask = mock(Task.class);
+
+        when(mockTask.isSuccessful()).thenReturn(true);
+        when(mockTask.getResult()).thenReturn(new FakeAuthResult(mockFirebaseUser));
+        when(mockActivityHelper.getApplicationContext()).thenReturn(mockContext);
+        when(mockActivityHelper.getFlowParams()).thenReturn(mockFlowParams);
+        credentialSignInHandler.onComplete(mockTask);
+
+        ArgumentCaptor<Intent> intentCaptor = ArgumentCaptor.forClass(Intent.class);
+        ArgumentCaptor<Integer> intCaptor = ArgumentCaptor.forClass(Integer.class);
+        verify(mockActivty).startActivityForResult(intentCaptor.capture(), intCaptor.capture());
+        Intent capturedIntent = intentCaptor.getValue();
+        assertEquals(RC_SAVE_CREDENTIALS, (int) intCaptor.getValue());
+        assertEquals(
+                SaveCredentialsActivity.class.getName(),
+                capturedIntent.getComponent().getClassName());
+        assertEquals(
+                TestConstants.EMAIL,
+                capturedIntent.getExtras().getString(ExtraConstants.EXTRA_EMAIL));
+        assertEquals(
+                TestConstants.NAME,
+                capturedIntent.getExtras().getString(ExtraConstants.EXTRA_NAME));
+        assertEquals(
+                TestConstants.PHOTO_URL,
+                capturedIntent.getExtras().getString(ExtraConstants.EXTRA_PROFILE_PICTURE_URI));
+    }
+
+    @Test
+    public void testSignInFailed_withFacebookAlreadyLinked() {
+        Activity mockActivty = mock(Activity.class);
+        ActivityHelper mockActivityHelper = mock(ActivityHelper.class);
+        FirebaseAuth mockFirebaseAuth = mock(FirebaseAuth.class);
+        IDPResponse mockIdpResponse = mock(IDPResponse.class);
+        CredentialSignInHandler credentialSignInHandler = new CredentialSignInHandler(
+                mockActivty,
+                mockActivityHelper,
+                RC_ACCOUNT_LINK,
+                RC_SAVE_CREDENTIALS,
+                mockIdpResponse);
+
+        Context mockContext = mock(Context.class);
+        Task mockTask = mock(Task.class);
+        FlowParameters mockFlowParams = mock(FlowParameters.class);
+        // pretend there was already an account with this email
+        when(mockTask.getException()).thenReturn(
+                new FirebaseAuthUserCollisionException(LINKING_ERROR, LINKING_EXPLANATION));
+        when(mockIdpResponse.getEmail()).thenReturn(TestConstants.EMAIL);
```

Comment:
Use a concrete instance of IDPResponse rather than a mock. The general principle behind this is "don't mock data" - IDPResponse is a simple object, so you should just be able to construct one and avoid cluttering the test with unnecessary mock expectations.